File: ios/ios/Views/Weather/DailyForecastView.swift
import SwiftUI
import shared

struct DailyForecastView: View {
    var weatherData: WeatherDataResult
    var body: some View {
        VStack {
            ForEach(1...weatherData.forecast.forecastDay.count, id: \.self) { item in
                DailyListItem(weatherData: weatherData.forecast.forecastDay[item - 1], isDay: 1)
            }
        }
        .padding(.top, 15)
    }
}

struct DailyListItem: View {
    let weatherData: ForecastDayWeatherObject
    let isDay: Int
    var body: some View {
        ZStack(alignment: .leading) {
            Color("InterfaceGray")
                .cornerRadius(10)
                .opacity(0.5)
            HStack(alignment: .center) {
                Image(
                    getIconForCodeAndName(
                        isDay: Int32(isDay),
                        code: weatherData.dayDetails.condition.code
                    )
                )
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .frame(width: 30, height: 30)
                    .padding(.leading, 15)
                Text(parseWeekday(date: weatherData.date))
                    .font(.custom(ManropeRegular, size: 18))
                    .padding(.leading, 10)
                Spacer()
                HStack(alignment: .bottom) {
                    Text(String(Int(weatherData.dayDetails.highTempFahrenheit)) + "°")
                        .font(.custom(TexGyreHerosBold, size: 27))
                        .padding(.trailing, 0)
                    Text(String(Int(weatherData.dayDetails.lowTempFahrenheit)) + "°")
                        .font(.custom(TexGyreHerosBold, size: 20))
                        .colorMultiply(Color("AccentColor").opacity(0.5))
                        .padding(.trailing, 15)
                        .padding(.leading, 0)
                        .padding(.bottom, 1)
                }
            }
            .padding(.vertical, 10)
        }
        .padding(.horizontal, 20)
        .padding(.bottom, 8)
    }
}

func parseWeekday(date: String) -> String {
    let dateFormat = DateFormatter()
    dateFormat.dateFormat = "yyyy-mm-dd"
    let parsedDate = dateFormat.date(from: date)!
    return dateFormat.weekdaySymbols[Calendar.current.component(.weekday, from: parsedDate) - 3] // subtract 3 to get correctly formatted weekday
}

func getCurrentTime() -> String {
    let dateFormatter = DateFormatter();
    dateFormatter.dateFormat = "h:mm a";
    return dateFormatter.string(from: Date())
}

File: android/src/main/java/com/thunderstorm/app/android/view/weather/DailyForecastView.kt
package com.thunderstorm.app.android.view.weather

import android.content.Context
import android.util.Log
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.LocalContentColor
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.thunderstorm.app.android.R
import com.thunderstorm.app.android.theme.TexGyreHeros
import com.thunderstorm.app.android.utils.getIconForNameAndCode
import com.thunderstorm.app.model.weather.WeatherDataResult
import com.thunderstorm.app.model.weather.forecast.ForecastDayWeatherObject
import java.text.SimpleDateFormat
import java.util.Locale
import kotlin.math.roundToInt

@Composable
fun DailyForecastView(
    weatherData: WeatherDataResult,
    context: Context
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 20.dp)
    ) {
        weatherData.forecast.forecastDay.forEach { item ->
            DailyListItem(
                weatherData = item,
                isDay = weatherData.current.isDay,
                context = context
            )
        }
    }
}

@Composable
fun DailyListItem(
    weatherData: ForecastDayWeatherObject,
    isDay: Int,
    context: Context
) {

    // Parse date so that SimpleDateFormat library actually gives us
    // the right date ._.
    val parsedDate = weatherData.hourDetails[0].localTime.let { details ->
        val dayOfMonth = details.substring(8, 10)
        val parsedDay = dayOfMonth.toInt() - 2
        details.replace(dayOfMonth, parsedDay.toString())
    }

    val weekdayFormat = SimpleDateFormat("EEEE", Locale.getDefault())
    val weatherServiceFormat = SimpleDateFormat("yyyy-mm-dd hh:mm", Locale.getDefault())
    val weekdayParsed = weekdayFormat.format(weatherServiceFormat.parse(parsedDate)!!)
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 10.dp, start = 20.dp, end = 20.dp)
            .clip(RoundedCornerShape(10.dp))
            .background(colorResource(id = R.color.interface_gray).copy(0.5F)),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically
        ) {
            Image(
                painter = painterResource(
                    id = context.getIconForNameAndCode(
                        isDay,
                        weatherData.dayDetails.condition.code
                    )
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(start = 15.dp, top = 15.dp, bottom = 15.dp)
                    .size(30.dp)
            )
            Text(
                text = weekdayParsed,
                modifier = Modifier.padding(start = 15.dp)
            )
        }
        Row(
            modifier = Modifier.padding(end = 20.dp)
        ) {
            weatherData.dayDetails.let { data ->
                Text(
                    text = buildAnnotatedString {
                        withStyle(
                            SpanStyle(
                                fontSize = 25.sp,
                                fontWeight = FontWeight.Bold,
                                fontFamily = TexGyreHeros
                            )
                        ) {
                            append("""${data.highTempFahrenheit.roundToInt()}° """)
                        }
                        withStyle(
                            SpanStyle(
                                fontSize = 18.sp,
                                fontWeight = FontWeight.Bold,
                                fontFamily = TexGyreHeros,
                                color = LocalContentColor.current.copy(0.7F)
                            )
                        ) {
                            append("""${data.lowTempFahrenheit.roundToInt()}°""")
                        }
                    }
                )
            }
        }
    }
}